Userland/Libraries/LibWeb/StorageAPI/NavigatorStorage.cpp

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2024, Jamie Mansfield <jmansfield@cadixdev.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJS/Runtime/Realm.h>
+#include <LibWeb/HTML/Scripting/Environments.h>
+#include <LibWeb/StorageAPI/NavigatorStorage.h>
+#include <LibWeb/StorageAPI/StorageManager.h>
+
+namespace Web::StorageAPI {
+
+NavigatorStorage::NavigatorStorage(HTML::EnvironmentSettingsObject& relevant_settings_object)
+    : m_relevant_settings_object(relevant_settings_object)
+{
+}
+
+// https://storage.spec.whatwg.org/#dom-navigatorstorage-storage
+JS::NonnullGCPtr<StorageManager> NavigatorStorage::storage()
+{
+    // The storage getter steps are to return this’s relevant settings object’s StorageManager object.
+    return m_relevant_settings_object->storage_manager();
```

Comment:
I don't think it's quite right to store the `EnvironmentSettingsObject`. There's an explicit AO for "this's relevant settings object".

So if we write the spec steps literally, we would have:
```c++
return HTML::relevant_settings_object(*this).storage_manager();
```

But of course, `this` is not a `JS::Object` because it's a mixin. What we do in other mixins is to define a virtual getter on the mixin that subclasses have to implement. See `this_impl` in LibWeb/HTML/WindowOrWorkerGlobalScope.h as an example.

So we would have:
```c++
class NavigatorStorage {

protected:
    virtual Bindings::PlatformObject const& this_navigator_storage_object() const = 0;

};
```

Which `Navigator` and `WorkerNavigator` would implement as simply:
```c++
private:
    virtual Bindings::PlatformObject const& this_navigator_storage_object() const override { return *this; }
```

And `NavigatorStorage::storage` would become:
```c++
JS::NonnullGCPtr<StorageManager> NavigatorStorage::storage()
{
    // The storage getter steps are to return this’s relevant settings object’s StorageManager object.
    return HTML::relevant_settings_object(this_navigator_storage_object()).storage_manager();
}
```

Reply:
This was really helpful, thank you :) I really appreciate the detailed notes you wrote.

It's much cleaner than what I wrote.